### PR TITLE
Get rid of TExecutionContext#inputTypeAt

### PR DESCRIPTION
**Pull request by hchandl**
TScalarBase had to change to a little, but the input types should be correct now and inputTypeAt shouldn't be necessary.

### CODE REVIEW

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/types/texpressions/TScalarBase.java [hunk 1]
```diff
@@ -143,6 +143,12 @@ public TPreptimeValue evaluateConstant(TPreptimeContext context, final LazyList<
             public ValueSource get(int i) {
                 TPreptimeValue ptValue = inputs.get(i);
                 ValueSource source = ptValue.value();
+                if (source == null) {
+                    source = ValueSources.getNullSource(ptValue.type());
+                }
+                else if (source.getType().typeClass() instanceof TString) {
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/types/texpressions/TScalarBase.java]: Why does every TString source need wrapped?

The addition above (null case) seems reasonable, given other recent investigations, but this line looks... odd.

**Comment by nathanlws** [comment]
One inline question about special casing of TString. Otherwise looks good.

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/types/texpressions/TScalarBase.java (hunk 1 above)
Comment (hchandl) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/types/texpressions/TScalarBase.java]: TPreptimeValue#value() doesn't have collation specified -- the correct collation is in TPreptimeValue#type().

Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/types/texpressions/TScalarBase.java]: Yuck. But makes sense.

**Comment by nathanlws** [comment]
Approve

**Pull request merged**